### PR TITLE
feat(rule-tester): stricter rule test validations

### LINKED ISSUE #9016: Enhancement(rule-tester): Stricter rule test validations
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

rule-tester

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Splitting out of #8211 -> #8969: this issue tracks any updates we'll need corresponding to https://github.com/eslint/eslint/pull/17654.

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by Vinccool96**
## PR Checklist

- [X] Addresses an existing open issue: fixes #9016 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Brings in changes from https://github.com/eslint/eslint/pull/17654. This is pretty much a copy & paste.

### CODE REVIEW

File: packages/rule-tester/src/RuleTester.ts [hunk 1]
```diff
@@ -929,130 +941,143 @@ export class RuleTester extends TestFramework {
...
-                      expectedSuggestion.data,
+                error.suggestions!.forEach(
+                  (expectedSuggestion: SuggestionOutput<MessageIds>, index) => {
+                    assert.ok(
+                      typeof expectedSuggestion === 'object' &&
+                        expectedSuggestion !== null,
+                      "Test suggestion in 'suggestions' array must be an object.",
+                    );
+                    // @ts-expect-error -- we purposely don't define `desc` on our types as the current standard is `messageId`
+                    const expectedDesc = expectedSuggestion.desc as string;
```
Comment (auvred) [inline packages/rule-tester/src/RuleTester.ts:986]: These two lines used to be in the `if (hasOwnProperty(expectedSuggestion, 'desc')) { ...`, but they have now been moved out of that block. I'm note sure if this change is related to this PR. WDYT?

File: packages/rule-tester/src/RuleTester.ts [hunk 2]
```diff
@@ -929,130 +941,143 @@ export class RuleTester extends TestFramework {
...
-                      actualSuggestion.desc,
-                      rule.meta.messages[expectedSuggestion.messageId],
-                      expectedSuggestion.data,
-                    );
 
-                  assert.ok(
-                    unsubstitutedPlaceholders.length === 0,
-                    `The message of the suggestion has ${unsubstitutedPlaceholders.length > 1 ? `unsubstituted placeholders: ${unsubstitutedPlaceholders.map(name => `'${name}'`).join(', ')}` : `an unsubstituted placeholder '${unsubstitutedPlaceholders[0]}'`}. Please provide the missing ${unsubstitutedPlaceholders.length > 1 ? 'values' : 'value'} via the 'data' property for the suggestion in the context.report() call.`,
-                  );
-
```
Comment (auvred) [inline packages/rule-tester/src/RuleTester.ts:1021]: These lines were added 4 days ago in the https://github.com/typescript-eslint/typescript-eslint/pull/9039 :smile: Could you revert them, please?

File: packages/rule-tester/tests/eslint-base/eslint-base.test.js [hunk 3]
```diff
@@ -473,7 +473,7 @@ describe("RuleTester", () => {
                     "bar = baz;"
                 ],
                 invalid: [
-                    { code: "var foo = bar; var baz = quux", errors: [{ type: "VariableDeclaration" }, null] }
+                  { code: "var foo = bar; var baz = quux", errors: [{ message: "Bad var.", type: "VariableDeclaration" }, null] }
```
Comment (auvred) [inline packages/rule-tester/tests/eslint-base/eslint-base.test.js:476]:
```suggestion
                    { code: "var foo = bar; var baz = quux", errors: [{ message: "Bad var.", type: "VariableDeclaration" }, null] }
```

We should try to keep this file 1:1 with the upstream tests, even in formatting!

**Review by auvred — CHANGES_REQUESTED** [review]

File: packages/rule-tester/tests/eslint-base/eslint-base.test.js (hunk 3 above)
Comment (Vinccool96) [reply packages/rule-tester/tests/eslint-base/eslint-base.test.js]: Done

**Comment by Vinccool96** [comment]
I'm gonna need some help for the failing tests

**Comment by Vinccool96** [comment]
@auvred Is everything good to merge?

File: packages/rule-tester/src/RuleTester.ts [hunk 4]
```diff
@@ -929,130 +940,161 @@ export class RuleTester extends TestFramework {
...
+              : Boolean(error.suggestions);
+            const hasSuggestions = message.suggestions !== void 0;
+            // @ts-expect-error -- we purposely don't verify for undefined
+            const messageSuggestions: Linter.LintSuggestion[] =
+              message.suggestions;
+
+            if (!hasSuggestions && expectsSuggestions) {
+              assert.ok(
+                !error.suggestions,
+                `Error should have suggestions on error with message: "${message.message}"`,
```
Comment (auvred) [inline packages/rule-tester/src/RuleTester.ts:960]:
```suggestion
            if (message.suggestions === void 0 && expectsSuggestions) {
              assert.ok(
                !error.suggestions,
                `Error should have suggestions on error with message: "${message.message}"`,
```

[Refactor] `hasSuggestions` is only used twice, perhaps we can inline a `message.suggestions !== void 0` check in the if conditions, that way TypeScript will understand that it is safe to use `message.suggestions` without putting `// @ts-expect-error`?

File: packages/rule-tester/src/RuleTester.ts [hunk 5]
```diff
@@ -929,130 +940,161 @@ export class RuleTester extends TestFramework {
...
+                    const unsubstitutedPlaceholders =
+                      getUnsubstitutedMessagePlaceholders(
+                        actualSuggestion.desc,
+                        rule.meta.messages[expectedSuggestion.messageId],
+                        expectedSuggestion.data,
+                      );
+
+                    assert.ok(
+                      unsubstitutedPlaceholders.length === 0,
+                      `The message of the suggestion has ${unsubstitutedPlaceholders.length > 1 ? `unsubstituted placeholders: ${unsubstitutedPlaceholders.map(name => `'${name}'`).join(', ')}` : `an unsubstituted placeholder '${unsubstitutedPlaceholders[0]}'`}. Please provide the missing ${unsubstitutedPlaceholders.length > 1 ? 'values' : 'value'} via the 'data' property for the suggestion in the context.report() call.`,
```
Comment (auvred) [inline packages/rule-tester/src/RuleTester.ts:1008]: [Refactor] This change doesn't seem to be related to the upstream PR. Is there any reason to move this peace of code (`unsubstitutedPlaceholders` + `assert.ok(unsubstitutedPlaceholders.length === 0, ...`) before the ` if (hasOwnProperty(expectedSuggestion, 'desc')) {` line?

File: packages/rule-tester/tests/eslint-base/eslint-base.test.js [hunk 6]
```diff
@@ -913,14 +944,28 @@ describe("RuleTester", () => {
...
-            valid: ["Eval(foo)"],
-            invalid: [{
-                code: "eval(foo)",
-                errors: [{}]
-            }]
+          valid: ["Eval(foo)"],
+          invalid: [{
+            code: "eval(foo)",
+            errors: [{}]
+          }]
```
Comment (auvred) [inline packages/rule-tester/tests/eslint-base/eslint-base.test.js:954]: [Refactor] The upstream uses 4 spaces for indentation. Could you please copy-paste it without re-indenting it? (we try to keep this file as close to the upstream as possible :slightly_smiling_face:)

File: packages/rule-tester/tests/eslint-base/eslint-base.test.js [hunk 7]
```diff
@@ -1048,7 +1093,7 @@ describe("RuleTester", () => {
                 {
                     code: "eval(foo)",
                     parser: require.resolve("esprima"),
-                    errors: [{ line: 1 }]
+                  errors: [{ message: "eval sucks.", line: 1 }]
```
Comment (auvred) [inline packages/rule-tester/tests/eslint-base/eslint-base.test.js:1096]: Same as above ^

**Review by auvred — CHANGES_REQUESTED** [review]
Left a few minor requests but overall looks great!

---

Btw, there is no need to ping individuals for review. We have a queue and we get to things when we can. https://typescript-eslint.io/contributing/pull-requests

**Comment by Vinccool96** [comment]
@auvred I know I'm not supposed to ping, but the Header Rules check keeps getting stuck, so the PR never shows up in the queue. Is there any way to mark it as successful?

File: packages/rule-tester/src/RuleTester.ts [hunk 8]
```diff
@@ -871,18 +885,6 @@ export class RuleTester extends TestFramework {
...
-                message.message,
-                rule.meta.messages[message.messageId],
-                error.data,
-              );
-
-            assert.ok(
-              unsubstitutedPlaceholders.length === 0,
-              `The reported message has ${unsubstitutedPlaceholders.length > 1 ? `unsubstituted placeholders: ${unsubstitutedPlaceholders.map(name => `'${name}'`).join(', ')}` : `an unsubstituted placeholder '${unsubstitutedPlaceholders[0]}'`}. Please provide the missing ${unsubstitutedPlaceholders.length > 1 ? 'values' : 'value'} via the 'data' property in the context.report() call.`,
-            );
-
```
Comment (auvred) [inline packages/rule-tester/src/RuleTester.ts:885]: Where did the `assert.ok(unsubstitutedPlaceholders.length === 0, ...` assertion go? All I see is that it has been deleted, but not restored elsewhere..

File: packages/rule-tester/tests/eslint-base/eslint-base.test.js [hunk 9]
```diff
@@ -2087,85 +2206,76 @@ describe("RuleTester", () => {
             }, /A fatal parsing error occurred in suggestion fix\.\nError: .+\nSuggestion output:\n.+/u);
         });
 
-        it("should throw if the suggestion description doesn't match", () => {
+        it("should fail when the suggestion property is neither a number nor an array", () => {
             assert.throws(() => {
```
Comment (auvred) [inline packages/rule-tester/tests/eslint-base/eslint-base.test.js:2210]: In the original pr, the case `"should fail when the suggestion property is neither a number nor an array"` was inserted before the case `"should throw if suggestion fix made a syntax error."`. Could you please do the same? Perhaps this would also improve the diff below

**Review by auvred — CHANGES_REQUESTED** [review]
I was a bit busy last month, that's why it took so long to re-review this PR :grimacing: 

I think the last blocker is the missing `assert.ok(unsubstitutedPlaceholders.length === 0, ...`!

---

Btw, could you please avoid force pushes in the future? It's harder for reviewer to understand what commits have been made since the last review

https://typescript-eslint.io/contributing/pull-requests

```
Please don't:

    Force push after opening a PR
        Reasoning: GitHub is not able to track changes across force pushes, which makes it take longer for us to perform incremental reviews
```

**Review by auvred — CHANGES_REQUESTED** [review]
Did you check this comment? https://github.com/typescript-eslint/typescript-eslint/pull/9086#discussion_r1656976881

We still miss this assertion

**Comment by Vinccool96** [comment]
The assertion is here https://github.com/Vinccool96/typescript-eslint/blob/4eebf7df8d89a919fbc03b69c9405d118e9d066a/packages/rule-tester/src/RuleTester.ts#L1047-L1057

**Comment by auvred** [comment]
On `main` there are two different assertions:

https://github.com/typescript-eslint/typescript-eslint/blob/de18a6ee443f591248dc04897fc5279675243ba1/packages/rule-tester/src/RuleTester.ts#L874-L884

https://github.com/typescript-eslint/typescript-eslint/blob/de18a6ee443f591248dc04897fc5279675243ba1/packages/rule-tester/src/RuleTester.ts#L1032-L1042

**Comment by Vinccool96** [comment]
Alright, let me plop it back in

**Comment by Vinccool96** [comment]
The prodigal son has returned

**Review by auvred — DISMISSED** [review]
Thanks for working on this!

**Review by JoshuaKGoldberg — APPROVED** [review]
The tests for this looked like a real slog, thanks so much for doing this! 🙌 

Just a couple of teeny suggestions. Nothing major from me. We can apply them before merging if you don't get a chance before our Monday release.

**Review by JoshuaKGoldberg — DISMISSED** [review]
Thanks! 🙌

**Comment by Vinccool96** [comment]
Had to merge since the main branch had tests that would've failed with the stricter rule tester (tests added by me, what was I thinking!)

**Review by auvred — APPROVED** [review]
